front/src/HOC/withLogger.test.js

```diff
@@ -0,0 +1,211 @@
+        this.props.log('inner component updated');
+      }
+    }
+    componentWillUnmount() {
+      if (this.props.log) {
+        this.props.log('inner component will unmount');
+      }
+    }
+    render() {
+      return <></>;
+    }
+  };
+  ComponentForTest.propsTypes = {
+    log: PropTypes.func,
+  };
+  ComponentForTest.defaultProps = {
+    log: () => {},
+  };
+  return ComponentForTest;
+};
+const fakeLogger = jest.fn(data => data);
+
+describe('Testing withLogger', () => {
+  afterEach(() => {
+    fakeLogger.mockClear();
+  });
+  test('should add log function', () => {
+    const value = withLogger();
+    expect(value).toBeFunction;
+  });
+  test('should add log function with specified params', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: [
+        'didMount',
+        'shouldUpdate',
+        'didUpdate',
+        'willUnmount',
+      ],
+    })(TestComponent);
+    const { rerender, unmount } = render(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    rerender(<WrappedComponent qw="1" />);
+    expect(fakeLogger).toHaveBeenCalledTimes(3);
+    unmount(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(4);
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      1,
+      'component mounted'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      2,
+      'component should update'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      3,
+      'component updated'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      4,
+      'component will unmount'
+    );
+  });
+  test('should add log function with specified HOC name', () => {
+    const WrappedComponent = withLogger()(TestComponent);
+    expect(WrappedComponent.displayName).toBe(
+      'withLogger(TestComponent)'
+    );
+  });
+  test('should log specified lifecycles', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: ['didMount', 'willUnmount'],
+    })(TestComponent);
+    const { unmount } = render(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    unmount(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(2);
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      1,
+      'component mounted'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      2,
+      'component will unmount'
+    );
+  });
+  test('should log props', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: ['props'],
+    })(TestComponent);
+    const {} = render(<WrappedComponent q="1" w="erty" />);
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    expect(fakeLogger).toHaveBeenNthCalledWith(1, {
+      q: '1',
+      w: 'erty',
+    });
+  });
+  test('should log props first', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: [
+        'props',
```

Comment:
I'm a bit confused here. Why do we have props in the list of lifecycle method? Could you please add more details about this idea?

Reply:
Yeah, seems it'll be better to use another object field for this param.
The idea was that it works in the same way as for lifecycles.